### ISSUE TITLE
keeps asking for session key

### ISSUE THREAD

**Issue by Joshi425** [raiyni/bitwarden-plugin#8]
hello there, 
I'd love to use this plugin but I can't get it to work on my win 11 maschine.

everytime I enable it it says: "Error loading vault. Try logging out of Bitwarden CLI and loggin back in" and then prompts me for a Session Key. when I paste a valid one it just prompts me again with the same error.

`bw list items --search runescape.com --session xxxx`

works in Powershell.
I'm running Windows 11
and bw-cli 1.22.0
my bitwarden server is a selfhosted vaultwarden, but I don't think this is an issue since bitwarden-cli is the same

I tried to save the BW_SESSION as USER and SYSTEM env var but no Luck. Can you please help me get this working or point into a direction?

**Comment by FairZ33** [comment]
Hello, I am having this same problem. Windows 10, bw-cli 1.22.1. Would love to have this work again.

**Comment by raiyni** [comment]
According to this issue, there seems to be problems with 1.21+ https://github.com/raiyni/bitwarden-plugin/issues/6

I haven't been able to replicate these issues

**Comment by FairZ33** [comment]
I just downgraded to 1.20.0 and I continue to get the prompt. Maybe I am setting it up wrong. I've placed bw.exe in "%userprofile%\Documents" and updated the path in the plug-in as "C:\Users\username\Documents" I logged into bw using `login --raw` copied the resulting session key and pasted it into a user environment variable "BW_SESSION". I also created a matching system variable just in case.

When launching RuneLite I get the error "Error loading vault. Try logging out of Bitwarden CLI and logging back in" which I have tried and updated the key.

I've tried manually providing the session key and still get the error. Like the other user above, running `bw list items --search runescape.com --session xxxx` returns the vault without issue. 

I have also tried deleting "%appdata%\Bitwarden\data.json" and "%appdata%\BitwardenCLI\data.json" and starting from scratch. 

![image](https://user-images.githubusercontent.com/20650533/169391925-892b92f4-3759-4c24-9bf9-960249ed0265.png)
![image](https://user-images.githubusercontent.com/20650533/169392394-f4054ba4-ed77-4e30-b171-be6fbedc9717.png)
![image](https://user-images.githubusercontent.com/20650533/169390422-91ca9cc1-978c-4199-912d-0d3ae9758167.png)
![image](https://user-images.githubusercontent.com/20650533/169389637-404947a1-6e84-426d-9f7a-d94ff252c9d3.png)
![image](https://user-images.githubusercontent.com/20650533/169390291-35dbc5f4-6361-4e92-bf6c-a0f225690e90.png)
![image](https://user-images.githubusercontent.com/20650533/169390162-139d10be-40cf-4eac-ab4d-d806e6f7e4a6.png)
![image](https://user-images.githubusercontent.com/20650533/169390874-3f32a6be-b9bd-4936-be7a-480fc59251b9.png)

**Comment by raiyni** [comment]
Remove the tick marks from the config option and add \bw to the end of it. 

The instructions on the readme meant for you to add that value to your system path variable in the last image, not the config. But setting it in config works too if it's right

**Comment by FairZ33** [comment]
> Remove the tick marks from the config option and add \bw to the end of it.
> 
> The instructions on the readme meant for you to add that value to your system path variable in the last image, not the config. But setting it in config works too if it's right

Brilliant. It's working now. I did a couple of things:

1. I updated the path in runelite to `C:\users\username\documents\bw` as suggested, adding `\bw` and removing the quotes around the path. 
2. I was now getting an error "Your session key may be wrong". I had quotes around the session key in the user environment variable value field. I removed those and tried again... it worked!

I'd love to see a screenshot of how the system path env variable is supposed to be setup, if you don't mind!

**Comment by raiyni** [comment]
Glad to hear. I can't do a screenshot right now but you should be able to double click on the PATH variable if you bring up the Edit System Environment Variables option. Then you would add a line with `%USERPROFILE%\Documents` and restart the client. Then you could reset the config option

**Comment by ppyLEK** [comment]
I am late to this party but I had the same problem so incase someone else is having this problem here is something you can try: Downgrade to 1.22.0 and edit the path in plugin to include "" symbols even if posts earlier say you should not have those.

**Comment by nathlrowe** [comment]
This issue still applies in 2023. My session key is correct, my "bw location" path is correct (doesn't have quotes, ends in "...\bw"), and the session key is correct and valid. I'm trying to login with the "Active" method. I'm using version 2023.10.0 of the bitwarden CLI, latest version of the plugin. Still getting "your session key may be invalid".

**Comment by raiyni** [comment]
@nathlrowe what os version are you using?

**Comment by Rheon-D** [comment]
Just thought I would throw my 2cents into this cause I just had some fun.

I was using the latest CLI (2024.2.1) I would intermittently have issues logging in.
Either the plugin would never actually enter the password or would ask me for my session key endlessly. 

Opening the project up in intellij and logging the output showed me two things.

1. The latest version of the CLI would occasionally report **_mac failed_**.
 mac failed would intermittently show up if the same command or a bw status was run from inside the terminal.
 The output of the command would still **eventually** be successful and the expected result would be returned, but because the plugin explicitly handles any mention of "mac failed" as an error the password would not be entered on the login page. 
 
2. The plugin doesn't enjoy JsonNull if it occupies a value in either username or password of the returned search list. 
Like a bung login entry with no password set.
https://github.com/raiyni/bitwarden-plugin/blob/b60ea7ac62d5a4babe6a2139197d202ec3334e74/src/main/java/io/ryoung/bitwarden/Credential.java#L26-L30

TL;DR I downgraded my cli version (1.22.1) deleted my data.json and the plugin works again.

**Comment by raiyni** [comment]
> Just thought I would throw my 2cents into this cause I just had some fun.
> 
> I was using the latest CLI (2024.2.1) I would intermittently have issues logging in. Either the plugin would never actually enter the password or would ask me for my session key endlessly.
> 
> Opening the project up in intellij and logging the output showed me two things.
> 
>     1. The latest version of the CLI would occasionally report **_mac failed_**.
>        mac failed would intermittently show up if the same command or a bw status was run from inside the terminal.
>        The output of the command would still **eventually** be successful and the expected result would be returned, but because the plugin explicitly handles any mention of "mac failed" as an error the password would not be entered on the login page.
> 
>     2. The plugin doesn't enjoy JsonNull if it occupies a value in either username or password of the returned search list.
>        Like a bung login entry with no password set.
>        https://github.com/raiyni/bitwarden-plugin/blob/b60ea7ac62d5a4babe6a2139197d202ec3334e74/src/main/java/io/ryoung/bitwarden/Credential.java#L26-L30
> 
> 
> TL;DR I downgraded my cli version (1.22.1) deleted my data.json and the plugin works again.

I've updated  to `2024.2.1` and have yet to see this issue.  I'd be interested in a PR that would maybe give a retry button on the popup with some different error.

@Rheon-D What system and how do you have bw installed?

**Comment by Rheon-D** [comment]
Running Win11 x64, bw is thrown into a random folder and referenced via `PATH`

I've noticed today that the error I was having is more to do with me being an idiot than an issue with bw itself.

Briefly:
The issue seems to be caused when `BW_SESSION` is set to an invalid value and the vault is in its unlocked state.
The two different versions of bw seem to handle this state differently. 
The newer version spams a few `mac failed` messages into stderr.
The older version seems to only put one message, if any.
Because stdout and stderr are merged the newer cli will prepend `mac failed` before returning the result using the provided session key. 

Can probably sort this by not merging the outputs and instead checking if a response is received in stdout, because it seems bw will only return a result if successful, at least in this case. In other cases stdout is empty and stderr will have some output.
I'll look at this again when I get a chance, possibly do up a PR if you don't beat me to it.

**Comment by raiyni** [comment]
> Running Win11 x64, bw is thrown into a random folder and referenced via `PATH`
> 
> I've noticed today that the error I was having is more to do with me being an idiot than an issue with bw itself.
> 
> Briefly: The issue seems to be caused when `BW_SESSION` is set to an invalid value and the vault is in its unlocked state. The two different versions of bw seem to handle this state differently. The newer version spams a few `mac failed` messages into stderr. The older version seems to only put one message, if any. Because stdout and stderr are merged the newer cli will prepend `mac failed` before returning the result using the provided session key.
> 
> Can probably sort this by not merging the outputs and instead checking if a response is received in stdout, because it seems bw will only return a result if successful, at least in this case. In other cases stdout is empty and stderr will have some output. I'll look at this again when I get a chance, possibly do up a PR if you don't beat me to it.

Can you provide exact steps to reproduce this? When I fudge my session key I get `? Master password:` 

Which prompts `Vault is locked`. I think that message could be improved though.

**Comment by Rheon-D** [comment]
> Can you provide exact steps to reproduce this? When I fudge my session key I get `? Master password:`
> 
> Which prompts `Vault is locked`. I think that message could be improved though.

I think its important for it to be an old key, or something that matches a key.  I only get `? Master password:` if:
the vault is locked or `BW_SESSION` is set to garbage eg. 'foobarbang' 

For Terminal

- Unlock vault, set `BW_SESSION` to that key.
- Lock vault and close terminal
- Reopen terminal and unlock again
- Running `bw list items --search <whatever>` should show `mac failed\n? Master password:`
- Running the same command with `--session <correct-key>` should show `mac failed` followed by the expected jsonobject on the next line.

For RuneLite:

- Unlock vault, set `BW_SESSION` to that key. 
- Lock and unlock again, do not set `BW_SESSION`
- Start RuneLite. Plugin should prompt `Error loading vault` and ask for the session key.

If that happens, for myself at least I am stuck in a loop until I either update or delete `BW_SESSION` and restart RL.

You probably know this part but, If you test this in IJ you will have to restart IJ completely so it will grab the garbage env key and pass it to the jvm on test.

**Comment by raiyni** [comment]
> > Can you provide exact steps to reproduce this? When I fudge my session key I get `? Master password:`
> > Which prompts `Vault is locked`. I think that message could be improved though.
> 
> I think its important for it to be an old key, or something that matches a key. I only get `? Master password:` if: the vault is locked or `BW_SESSION` is set to garbage eg. 'foobarbang'
> 
> For Terminal
> 
>     * Unlock vault, set `BW_SESSION` to that key.
> 
>     * Lock vault and close terminal
> 
>     * Reopen terminal and unlock again
> 
>     * Running `bw list items --search <whatever>` should show `mac failed\n? Master password:`
> 
>     * Running the same command with `--session <correct-key>` should show `mac failed` followed by the expected jsonobject on the next line.
> 
> 
> For RuneLite:
> 
>     * Unlock vault, set `BW_SESSION` to that key.
> 
>     * Lock and unlock again, do not set `BW_SESSION`
> 
>     * Start RuneLite. Plugin should prompt `Error loading vault` and ask for the session key.
> 
> 
> If that happens, for myself at least I am stuck in a loop until I either update or delete `BW_SESSION` and restart RL.
> 
> You probably know this part but, If you test this in IJ you will have to restart IJ completely so it will grab the garbage env key and pass it to the jvm on test.

Okay I do see it those instructions, thank you.

**Comment by raiyni** [comment]
Note: I don't see the null value issue though.

**Comment by raiyni** [comment]
@Rheon-D I believe https://github.com/raiyni/bitwarden-plugin/commit/ccf3f305aa2301701f58c1c9d28c1fe63d5bab8a fixes your main issue. I'm not sure I really care about supporting the null username/password issue but I don't see it happening. That should be covered by the `Failed to parse` exception.
